### PR TITLE
fe: check constraint of types fix #309

### LINKED ISSUE #309: No errors are raised for abstract feature with unmet constraint
Example:
```
ex is
  a is
  b (Set a) is abstract
```
Even though Set has type constraint `hasEquals` and `a` does not implement `hasEquals` there is no error raised when trying to compile the example.

### PR DESCRIPTION
**Pull request by michaellilltokiwa**
fixes #309,  #383

### CODE REVIEW

**Comment by michaellilltokiwa** [comment]
Unfortunately it now reports incompatible types twice for calls that are called with an incompatible type:


image

**Comment by michaellilltokiwa** [comment]
@fridis Would be good to have some general feedback even though it is not working correctly yet.
Is the approach I took any good?

**Comment by michaellilltokiwa** [comment]
@maxteufel your "be noisy" PR already pays off here. it's even showing the ansi-colors. cheers.

**Comment by michaellilltokiwa** [comment]
This is how the error currently looks like. The none exact position is definitely unpleasant.
```
sam@debian /home/not_synced/fuzion (issue/309_check_constraints_of_types)$ fz tests/reg_issue309_check_constraints_of_types/issue309.fz

/home/not_synced/fuzion/tests/reg_issue309_check_constraints_of_types/issue309.fz:29:3: error 1: Incompatible type parameter
  b Set a is abstract          // 1. should flag an error: Incompatible type parameter
--^
formal type parameter 'E' with constraint 'hasEquals<Set.E>'
actual type parameter 'issue309.a'


/home/not_synced/fuzion/tests/reg_issue309_check_constraints_of_types/issue309.fz:31:3: error 2: Incompatible type parameter
  b(T type) Set T is abstract  // 2. should flag an error: Incompatible type parameter
--^
formal type parameter 'E' with constraint 'hasEquals<Set.E>'
actual type parameter 'issue309.b.T'


/home/not_synced/fuzion/tests/reg_issue309_check_constraints_of_types/issue309.fz:34:3: error 3: Incompatible type parameter
  d c a is abstract            // 3. should flag an error: Incompatible type parameter
--^
formal type parameter 'T' with constraint 'hasEquals<issue309.c.T>'
actual type parameter 'issue309.a'

3 errors.
```

File: src/dev/flang/ast/AbstractType.java [hunk 1]
```diff
@@ -995,6 +995,46 @@ public String asString()
...
+        // Check that generics match formal generic constraints
+        var fi = featureOfType().generics().list.iterator();
+        var gi = generics().iterator();
+        while (fi.hasNext() &&
+              gi.hasNext()    ) // NYI: handling of open generic arguments
+          {
+            var f = fi.next();
+            var g = gi.next();
+            g.checkConstraints(pos);
+            if(compareTo(f.constraint()) != 0) {
```
Comment (fridis) [inline src/dev/flang/ast/AbstractType.java:1020]: I think this is missing a `g.`: 
```
  if (g.compareTo(f.constraint()) != 0)
```
Also, pleas add a space between `if` and `(` and place `{` into a new line.

Comment (michaellilltokiwa) [reply src/dev/flang/ast/AbstractType.java]: @fridis I fixed the indentation. What I want to express here is that the constraint is not the same as the feature itself like in i.e. hasEquals:
```
hasEquals(HE (hasEquals HE).type) is
```

Comment (michaellilltokiwa) [reply src/dev/flang/ast/AbstractType.java]: @fridis I added a NYI comment regarding the possible infinite recursion

**Pull request merged**